frame/elections-phragmen/src/lib.rs

```diff
@@ -127,95 +127,88 @@ type NegativeImbalanceOf<T> =
 /// Helper functions for migrations of this module.
 pub mod migrations {
 	use super::*;
-	use frame_support::{migration::StorageKeyIterator, Twox64Concat};
-
-	/// Migrate from the old legacy voting bond (fixed) to the new one (per-vote dynamic).
-	///
-	/// Will only be triggered if storage version is V1.
-	pub fn migrate_voters_to_recorded_deposit<T: Trait>(old_deposit: BalanceOf<T>) -> Weight {
-		if <Module<T>>::pallet_storage_version() == StorageVersion::V1 {
-			let mut count = 0;
-			<StorageKeyIterator<T::AccountId, (BalanceOf<T>, Vec<T::AccountId>), Twox64Concat>>::new(
-				<Voting<T>>::module_prefix(),
-				b"Voting",
-			)
-			.for_each(|(who, (stake, votes))| {
-				// Insert a new value into the same location, thus no need to do `.drain()`.
-				let deposit = old_deposit;
-				let voter = Voter { votes, stake, deposit };
-				<Voting<T>>::insert(who, voter);
-				count += 1;
-			});
+	use frame_support::{
+		traits::{PalletVersion, GetPalletVersion},
+		Twox64Concat,
+		migration::StorageKeyIterator,
+	};
 
-			PalletStorageVersion::put(StorageVersion::V2RecordedDeposit);
-			frame_support::debug::info!(
-				"🏛 pallet-elections-phragmen: {} voters migrated to V2PerVoterDeposit.",
-				count,
-			);
-			Weight::max_value()
+	pub fn migrate_to_3_0_0<T: Trait>(old_voter_bond: BalanceOf<T>, old_candidacy_bond: BalanceOf<T>) -> Weight {
+		let current_version = <Module<T> as GetPalletVersion>::current_version();
+		let maybe_storage_version = <Module<T> as GetPalletVersion>::storage_version();
+
+		if current_version == PalletVersion::new(3, 0, 0) {
```

Comment:
This will disallow running this migration for future versions meaning that people that don't upgrade before the next change to the phragmen pallet will not be able to run this.
I think you should only match on the storage version here.

Reply:
Hmm good point, but I wonder if that can create any sort of footgun in an edge case where we accidentally execute a migration twice?

Reply:
Why I am worried about for example is if someone is trying to do multiple upgrades at once. In my restricted scheme, they cannot shoot themselves in the foot. In yours, I think they can?

Reply:
For that we could set the version that you're intending to migrate **to** after the migration was successful.

Reply:
Can we set the version? The version will always be set to what is noted in Cargo.toml after a runtime upgrade afaik.

Reply:
Let's elaborate: 

Imagine we have another migration from 3.0.0 -> 4.0.0 (that also depends on the current storage layout introduced in 3.0.0).  So the migration in current PR is migration `M3` and the one introduced hypothetically in the future is `M4`. Also, assume, once we add M4 and bump to 4.0.0 we keep M3 in the code base. 

Now, someone who last had 2.0.0 want to upgrade to 4.0.0. If both migrations are written as you suggest (only assert on storage version) they can use the crate in version 4.0.0  but apply M3. This will apply this migration for them, but will write 4.0.0. in their storage version, which means that they can never apply M4. 

I think it all boils down to keeping migrations in the code base. If we remove the migration introduced in this PR upon introducing the next migration, we are safe and I can do what you suggest. Else, I prefer keeping it strict with asserting on both versions.

Reply:
Ok, but this leads me to think we should not write the current version to storage when `on_runtime_upgrade` is present and instead leave that to the migration so we can avoid your scenario.
If the migration defines the version that it migrates to your scenario works out just fine. Every migration loads the version from storage, checks that it is the expected one, migrates, and then sets the version that it migrated to.
@bkchr If you agree with this scheme I think we might have to adjust the `on_runtime_upgrade`.

Reply:
@kianenigma Your idea is that people pulling in this migration will either use it as is (in the context of a v3 pallet) or will modify it to fit their context (e.g. in a v4 pallet)?

Reply:
> Ok, but this leads me to think we should not write the current version to storage when on_runtime_upgrade is present and instead leave that to the migration so we can avoid your scenario.


Maybe `on_runtime_upgrade` of each module can return `Option<PalletVersion>`, and you get where I am going with this.

But this makes things complicated. I am fine with keeping the current scheme. The only hassle now is for external teams: if you are 3 major versions (i.e. 3 migrations behind) you cannot do them all at once. You need to do them one by one. But this hassle will protect them actually, so it is good.

Reply:
why not write version 3_0_0 in storage when this migration is successful, and remove this if condition ?

I would rather keep writing after on_runtime_upgrade automatically, this prevents more errors that it causes.

Reply:
I see that the main dilemma here is: 

1. Should this migration be executed as a `fn on_runtime_upgrade()` inside the pallet? 
2. Should this migration be executed as an additional upgrade directly given to executive? 

I assumed that in all cases the automatic write happens, while it is not the case. Automatic write happens only in case 1. 

Based on the assumption that the upgrade will be called from executive I can remove the additional `if` and fo a manual write.

Reply:
Honestly though, I think this migration should live inside the `fn on_runtime_upgrade()` of the pallet, and have all the needed if-conditions to make sure it is executed only when it is a jump from 2_0_0 to 3_0_0. Given this, the code can always live in substrate and allows chains that are multiple versions behind to also update them one by one.

Reply:
yes ideally it is better for it to live inside the on_runtime_upgrade, but how can we give the information about old deposits ?
Maybe as associated type on Config ? then this is bloating Config :-/ so I don't know what is best.

OnRuntimeUpgrade executes first CustomOnRuntimeUpgrade define at runtime level then pallets OnRuntimeUpgrade which writes into the storage the pallet version after their execution.

Reply:
> Maybe as associated type on Config ? then this is bloating Config :-/ so I don't know what is best.

Even worse about this is that now we bump to 3_0_0 because the change is breaking, and if we add new configs just for migration, we only can clean them if we bump to 4_0_0 because the removal itself will be a breaking change (I am doing my best here to adhere to semver, which is kinda hard to do in pallets.)

Reply:
> I see that the main dilemma here is:
> 
>     1. Should this migration be executed as a `fn on_runtime_upgrade()` inside the pallet?
> 
>     2. Should this migration be executed as an additional upgrade directly given to executive?
> 
> 
> I assumed that in all cases the automatic write happens, while it is not the case. Automatic write happens only in case 1.

I don't think that's right. I thought [this code](https://github.com/paritytech/substrate/blob/f884296f7436916909025f8b43c4bbf3e60e4c60/frame/support/src/dispatch.rs#L1349-L1358) means that the version is set for all pallets.